Ensure that SourceFiles always have a backing buffer in the SourceManager

DougGregor opened the pull request:
The "buffer ID" in a SourceFile, which is used to find the source file's contents in the SourceManager, has always been optional. However, the effectively every SourceFile actually does have a buffer ID, and the vast majority of accesses to this information dereference the optional without checking.

Update the handful of call sites that provided `nullopt` as the buffer ID to provide a proper buffer instead. These were mostly unit tests and testing programs, with a few places that passed a never-empty optional through to the SourceFile constructor. The code that created a source buffer to parse C `swift_attr` attributes also failed to pass that source buffer through.

Then, remove optionality from the representation and accessors. It is now the case that every SourceFile has a buffer ID, simplying a bunch of code.

DougGregor committed:
Provide the buffer ID for source files created for C swift_attr attributes

When rendering a swift_attr attribute for parsing, we were creating both
the backing buffer and a source file, but not providing the buffer's ID
to the source file, so we couldn't find the source code again. Fix that.

While here, also register the source file with the module, so we can
find the source file again based on location.

DougGregor committed:
Ensure that SourceFiles always have a backing buffer in the SourceManager

The "buffer ID" in a SourceFile, which is used to find the source file's
contents in the SourceManager, has always been optional. However, the
effectively every SourceFile actually does have a buffer ID, and the
vast majority of accesses to this information dereference the optional
without checking.

Update the handful of call sites that provided `nullopt` as the buffer
ID to provide a proper buffer instead. These were mostly unit tests
and testing programs, with a few places that passed a never-empty
optional through to the SourceFile constructor.

Then, remove optionality from the representation and accessors. It is
now the case that every SourceFile has a buffer ID, simplying a bunch
of code.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

CodaFi commented on lib/AST/Module.cpp:
```diff
@@ -3433,9 +3430,9 @@ ModuleDecl::computeFileIDMap(bool shouldDiagnose) const {
-    : FileUnit(FileUnitKind::Source, M), BufferID(bufferID ? *bufferID : -1),
+    : FileUnit(FileUnitKind::Source, M), BufferID(bufferID),
       ParsingOpts(parsingOpts), IsPrimary(isPrimary), Kind(K) {
   M.getASTContext().addDestructorCleanup(*this);
```
Should this assert that the buffer ID isn't `~0` or is that now valid (if... very unlikely to encounter in the wild)?

DougGregor replied:
Good idea!

DougGregor committed:
Ensure that we don't get a very-likely-wrong buffer ID

-1 was once used as a sentinel value for "no buffer"; make sure it
doesn't persist anywhere.

DougGregor commented:
https://github.com/swiftlang/llvm-project/pull/9269

@swift-ci please test

DougGregor commented:
https://github.com/swiftlang/llvm-project/pull/9269

@swift-ci please smoke test

DougGregor committed:
Silence warning about mismatched signs
